iothub_client/src/uamqp_messaging.c

```diff
@@ -430,46 +434,102 @@ int message_create_uamqp_encoding_from_iothub_message(IOTHUB_MESSAGE_HANDLE mess
+
+                if (value_type == AMQP_TYPE_STRING)
+                {
+                    // Codes_SRS_UAMQP_MESSAGING_09_014: [The message-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_string(uamqp_message_property, &string_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'message-id' property (string)");
+                        result = __FAILURE__;
+                    }
+                }
+                else if (value_type == AMQP_TYPE_ULONG)
+                {
+                    uint64_t ulong_value;
+
+                    // Codes_SRS_UAMQP_MESSAGING_09_014: [The message-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_ulong(uamqp_message_property, &ulong_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'message-id' property (ulong)");
+                        result = __FAILURE__;
+                    }
+                    else
+                    {
+                        (void)sprintf(string_buffer, "%llu", ulong_value);
+                        string_value = string_buffer;
+                    }
+                }
+                else if (value_type == AMQP_TYPE_UUID)
+                {
+                    uuid uuid_value;
+
+                    // Codes_SRS_UAMQP_MESSAGING_09_014: [The message-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_uuid(uamqp_message_property, &uuid_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'message-id' property (UUID)");
+                        result = __FAILURE__;
+                    }
+                    else if ((string_value = UUID_to_string((UUID*)uuid_value)) == NULL)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_015: [If message-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get the string representation of 'message-id' UUID");
+                        result = __FAILURE__;
+                    }
+                    else
+                    {
+                        (void)memcpy(string_buffer, string_value, sizeof(string_value));
```

Comment:
Note though my comment RE breaking this up into sub functions, this may not be as relevant(?) since I wrote before the feedback at head of this fcn.

---
In reply to: [140559919](https://github.com/Azure/azure-iot-sdk-c/pull/237#discussion_r140559919) [](ancestors = 140559919)

Reply:
Makes sense. I'll take option B

Reply:
I'll take option A.

This is the logic:
When the message-id is:
STRING => string_value is a const pointer to a char sequence held by uamqp
ULONG => string_value is a pointer to string_buffer
UUID => string_value is a new string that we need to free after using.

So the if check you are proposing is not sufficient to define if we should free or not.

Reply:
Fixed to:
                        (void)memcpy(string_buffer, string_value, sizeof(char) * strlen(string_value));